### PR TITLE
Serving React SPA with Zio Http

### LINKED ISSUE #2292: Zio-http and Reactjs example project
**Is your feature request related to a problem? Please describe.**
I am new to zio, i want to develop full-stack web app with `zio-http` and `react.js`.
I couldn't figure out how to serve static files after calling `npm run build` in the react.js project.

**Describe the solution you'd like**
It would be nice if zio-http documentation includes an example simple full-stack app with serving files.

### PR DESCRIPTION
**Pull request by rishi-raj-jain**
Resolves #2292
/claim #2292

### CODE REVIEW

**Comment by ahmetgurbuz1** [comment]
`import zio._
import zio.http._
import zio.http.Server
import java.io.File

object Api extends ZIOAppDefault {
  private val buildDirectory: List[String] = List("ui", "build")
  private val staticDir: List[String]      = buildDirectory.:+("static")
  private val indexHtmlDir                 = buildDirectory.:+("index.html").mkString(File.separator)

  val app =
    Http.collectHttp[Request] {
      case Method.GET -> Root        => Http.fromFile(new File(indexHtmlDir))
      case Method.GET -> Root / "static" / content / file =>
        val path: List[String] = content match {
          case "js"    => staticDir.:+("js").:+(file)
          case "css"   => staticDir.:+("css").:+(file)
          case "media" => staticDir.:+("media").:+(file)
        }
        Http.fromFile(new File(path.mkString(File.separator)))
    }

  override val run: ZIO[Any, Throwable, Nothing] = Server
    .serve(app.withDefaultErrorResponse)
    .provide(Server.defaultWithPort(8080))
}
`

**Comment by scalactic** [comment]
```
import zio._
import zio.http._
import zio.http.Server
import java.io.File

object Api extends ZIOAppDefault {
  private val buildDirectory: List[String] = List("ui", "build")
  private val staticDir: List[String]      = buildDirectory.:+("static")
  private val indexHtmlDir                 = buildDirectory.:+("index.html").mkString(File.separator)

  val app =
    Http.collectHttp[Request] {
      case Method.GET -> Root        => Http.fromFile(new File(indexHtmlDir))
      case Method.GET -> Root / "static" / content / file =>
        val path: List[String] = content match {
          case "js"    => staticDir.:+("js").:+(file)
          case "css"   => staticDir.:+("css").:+(file)
          case "media" => staticDir.:+("media").:+(file)
        }
        Http.fromFile(new File(path.mkString(File.separator)))
    }

  override val run: ZIO[Any, Throwable, Nothing] = Server
    .serve(app.withDefaultErrorResponse)
    .provide(Server.defaultWithPort(8080))
}

```

**Comment by scalactic** [comment]
It was getting 404 on some endpoints, now it gets HTTP 200 in all requests.

**Comment by rishi-raj-jain** [comment]
@ahmetgurbuz1 Hey, updated with a general syntax.

**Comment by scalactic** [comment]
Added example complete app -> https://github.com/ahmetgurbuz1/scala-zio-reactjs

**Comment by rishi-raj-jain** [comment]
Hey  it shouldn't require that much effort,
just scaffold a create react app, and then create two files: api.scala with the content in my PR & create build.sbt with `libraryDependencies += "dev.zio" %% "zio-http" % "3.0.0-RC2"`

so the file structure would become:
root
- all react files and folders
- build.sbt
- api.scala
- build (by react build commad)

let sbt run be the magician afterwards

File: docs/examples/advanced/fullstack-with-react-apps [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    
+    // Create HTTP route
+    val app = Http.collectHttp[Request] {
+        case Method.GET -> Root / "api" / "hello" => Handler.text("Hello World!").toHttp
+
+        // Uses netty's capability to write file content to the Channel
+        // Content-type response headers are automatically identified and added
+        // Adds content-length header and does not use Chunked transfer encoding
+        case Method.GET -> Root => Http.fromFile(new File(s"$buildDirectory/index.html"))
+        case Method.GET -> "" /: file => Http.fromFile(new File(s"$buildDirectory/$file"))
```
Comment (guizmaii) [inline docs/examples/advanced/fullstack-with-react-apps:27]: Don't we need to protect against `..`? Someone asking for `../../aFile` could try to download system files, no?

Comment (scalactic) [reply docs/examples/advanced/fullstack-with-react-apps:27]: File secrets is unsafe in current logic, right?

Comment (guizmaii) [reply docs/examples/advanced/fullstack-with-react-apps:27]: What do you call "File secrets"?

Comment (scalactic) [reply docs/examples/advanced/fullstack-with-react-apps:27]: Mounted secrets in kubernetes/Docker apps

Comment (guizmaii) [reply docs/examples/advanced/fullstack-with-react-apps:27]: I'm not a 100% sure, someone would need to give a try, but I think that if we don't protect against the `..` file access, any file in the file system is accessible to someone having access to the API

IMO, it'd be great if zio-http would provide an helper to expose only a directory and it's sub-directories. IIRC, Play was providing such an helper. I think they were also calculating the files' hash/etag to work better with browsers cache mechanism. I don't remember the name of this helper.

Comment (rishi-raj-jain) [reply docs/examples/advanced/fullstack-with-react-apps:27]: @guizmaii Can you share in which URL would be able to go through ".."? That path from the URL won't translate as is in the app I believe. PFA a recording of the same.

https://github.com/zio/zio-http/assets/46300090/fd9c2ea0-5e49-4b3c-846f-1744b6a13481

Comment (guizmaii) [reply docs/examples/advanced/fullstack-with-react-apps:27]: @rishi-raj-jain Can you try with `curl`, please? Someone malicious will use a script to try to steal data, not a browser

Comment (rishi-raj-jain) [reply docs/examples/advanced/fullstack-with-react-apps:27]: PFA

https://github.com/zio/zio-http/assets/46300090/97b037f6-471a-4d8b-b3b5-b81a1a2eab4f

Comment (guizmaii) [reply docs/examples/advanced/fullstack-with-react-apps:27]: What does "PFA" means?

Comment (rishi-raj-jain) [reply docs/examples/advanced/fullstack-with-react-apps:27]: Hey guizmaii, oh actually I'm used to writing that. it's please find attached. I was just referring to the video that I attached

Comment (rishi-raj-jain) [reply docs/examples/advanced/fullstack-with-react-apps:27]: @guizmaii 

Everything looking good here?

Comment (guizmaii) [reply docs/examples/advanced/fullstack-with-react-apps:27]: @rishi-raj-jain The ticket #2292 asks for an example project combining React and zio-http.
I'm not working for Ziverge but I'm not sure you can claim the prize with just some documentation

**Comment by rishi-raj-jain** [comment]
Hey, where do I push the example?

**Comment by rishi-raj-jain** [comment]
@jdegoes where do I push the example? Or is the work here enough?

File: docs/examples/advanced/fullstack-with-react-apps (hunk 1 above)
Comment (rishi-raj-jain) [reply docs/examples/advanced/fullstack-with-react-apps:27]: @guizmaii 

That's what I've shown in the videos and the demo. But yeah a place to upload the project would be fine.

> just

Oops

**Comment by rishi-raj-jain** [comment]
@jdegoes where do I push the example? Or is the work here enough?

**Comment by rishi-raj-jain** [comment]
@amitksingh1490 @jdegoes where do I push the example? Or is the work here enough?

**Comment by rishi-raj-jain** [comment]
@adamgfraser @jdegoes Is the work here enough?

**Comment by rishi-raj-jain** [comment]
@adamgfraser @jdegoes Can you have a look at this? Let me know where do I push the example. Attached the zip as of now.
[react-scala.zip](https://github.com/zio/zio-http/files/12259548/react-scala.zip)

**Comment by guizmaii** [comment]
@rishi-raj-jain Just push your example project in a new sbt module. If they want you to move it, they'll just ask you to move it

**Comment by rishi-raj-jain** [comment]
@guizmaii I still don't understand what am I missing to let the team help me figure the right path.

**Comment by scalactic** [comment]
Honestly, i hesitate if this PR really deserves 75 $.
@jdegoes i wouldn't create this PR, if i knew you were going to put bounty on this.

**Comment by rishi-raj-jain** [comment]
I think it does deserve the amount set by @jdegoes.

This adds an example & documentation for it makes the DX smooth.

I look at the previous examples and solutions shared, and none of them could figure the general way to serve the build exactly (one was switching on the kind of files, but missed handling other file types that can/would be included with the build). Some of us had a concern of if paths are going to be read and internal file access got leaked.

Satisfying all that takes time and effort.

**Comment by rishi-raj-jain** [comment]
I think it does deserve the amount set by @jdegoes. Also, as I see it, I did the PR.

This adds an example & documentation for it makes the DX smooth.

I look at the previous examples and solutions shared, and none of them could figure the general way to serve the build exactly (one was switching on the kind of files, but missed handling other file types that can/would be included with the build). Some of us had a concern of if paths are going to be read and internal file access got leaked.

Satisfying all that takes time and effort.

**Comment by rishi-raj-jain** [comment]
I think it does deserve the amount set by @jdegoes. Also, as I see it, I did the PR.

This adds an example & documentation for it makes the DX smooth.

I look at the previous examples and solutions shared, and none of them could figure the general way to serve the build exactly (one was switching on the kind of files, but missed handling other file types that can/would be included with the build). Some of us had a concern of if paths are going to be read and internal file access got leaked.

Satisfying all that takes time and effort.

**Comment by guizmaii** [comment]
@rishi-raj-jain 
> I still don't understand what am I missing to let the team help me figure the right path.

They're just super busy. Please commit your example project. If they're not happy with where you put it, they'll just ask you to move it when they have some time to review your PR

**Comment by jdegoes** [comment]
Please push the example project into a new SBT module so we can ensure it's kept up-to-date with ZIO HTTP itself. Then we can do a final review and get this merged in!

**Comment by rishi-raj-jain** [comment]
@jdegoes 

Shall I push to https://github.com/zio/zio-quickstarts?

**Comment by jdegoes** [comment]
I would do it in this repository as a new sbt project, e.g. `zio-http-react-example`.

**Comment by rishi-raj-jain** [comment]
@jdegoes This is ready for your review.

**Comment by rishi-raj-jain** [comment]
@jdegoes This is ready for your review.

**Comment by rishi-raj-jain** [comment]
@jdegoes This is ready for your review.

**Comment by rishi-raj-jain** [comment]
@jdegoes This is ready for your review.

**Comment by rishi-raj-jain** [comment]
@jdegoes why is this closed? I added the example

**Comment by ccellado** [comment]
@jdegoes explanation requested

**Comment by jdegoes** [comment]
I don't remember why I closed this (usually I close with comment), but if it's re-opened, conflicts are fixed, and the build passes, let's get it merged in!

**Comment by rishi-raj-jain** [comment]
I can't seem to have the GitHub permissions to re-open this, may you? 

Doesn't show the Reopen PR button as below for me:
Screenshot 2023-11-22 at 8 06 18 AM

**Comment by vigoo** [comment]
image

**Comment by rishi-raj-jain** [comment]
Oh alright, I'll make another PR and link it here then.

**Comment by rishi-raj-jain** [comment]
Created https://github.com/zio/zio-http/pull/2529

**Pull request closed without merging**